src/samples/EvalClient/EvaluationService.cs

```diff
@@ -0,0 +1,211 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+using CsvHelper;
+using System.Globalization;
+using CsvHelper.Configuration;
+using Microsoft.Agents.CopilotStudio.Client;
+using Microsoft.Agents.Protocols.Primitives;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.AI;
+
+namespace EvalClient;
+
+/// <summary>
+/// This class is responsible for handling the Evaluation service and managing the conversation with the Copilot Studio hosted bot and Azure OpenAI.
+/// </summary>
+/// <param name="copilotClient">Connection Settings for connecting to Copilot Studio</param>
+
+internal class EvaluationService(CopilotClient copilotClient, IChatClient chatClient) : IHostedService
+{
+    /// <summary>
+    /// This is the main thread loop that manages the back and forth communication with the Copilot Studio Bot. 
+    /// </summary>
+    /// <param name="cancellationToken"></param>
+    /// <returns></returns>
+    /// <exception cref="InvalidOperationException"></exception>
+    
+    static string inputEvalDataset = "./Data/Evaluation Dataset.csv";
+    static string outputEvalDataset = "./Data/Evaluation Dataset Results.csv";
+    
+    public async Task StartAsync(CancellationToken cancellationToken)
+    {
+        // Load the input file and parse the data.
+        var evalDataset = LoadCsvData(inputEvalDataset);
+
+        Console.WriteLine("\nRunning evaluation on Agent");
+        // Attempt to connect to the copilot studio hosted bot here
+        // if successful, this will loop though all events that the Copilot Studio bot sends to the client setup the conversation.
+        await foreach (Activity act in copilotClient.StartConversationAsync(emitStartConversationEvent: true, cancellationToken: cancellationToken))
+        {
+            if (act is null)
+            {
+                throw new InvalidOperationException("Activity is null");
+            }
+            // for each response, report to the UX
+            GetActivity(act);
+        }
+
+        // Once we are connected and have initiated the conversation, begin the message loop with the Console.
+        for (int count = 0; count < evalDataset.Count(); count++)
+        {
+            Console.WriteLine($"Evaluating dataset {count + 1}/{evalDataset.Count()}...");
+            var question = evalDataset[count].TestUtterance;
+
+            // Send the evaluation question to the Copilot Studio bot and await the response.
+            string response = String.Empty;
+            await foreach (Activity act in copilotClient.AskQuestionAsync(question, null, cancellationToken))
+            {
+                response += GetActivity(act);
```

Comment:
The GetActivity method can return EmptyString "" when the activityType is not "Message". Might be better to run the Evaluation conditionally when there is a non-empty response.

Reply:
Good point. Improved.